source/common/http/filter_manager.cc

```diff
@@ -1448,6 +1495,68 @@ void FilterManager::maybeEndEncode(bool end_stream) {
   if (end_stream) {
     ASSERT(!state_.encoder_filter_chain_complete_);
     state_.encoder_filter_chain_complete_ = true;
+    if (filter_manager_callbacks_.isHalfCloseEnabled()) {
+      // If independent half close is enabled the stream is closed when both decoder and encoder
+      // filter chains has completed or were aborted.
+      checkAndCloseStreamIfFullyClosed();
+    } else {
+      // Otherwise encoding end_stream always closes the stream (and resets it if request was not
+      // complete yet).
+      filter_manager_callbacks_.endStream();
+    }
+  }
+}
+
+void FilterManager::maybeEndDecode(bool terminal_filter_decoded_end_stream) {
+  if (terminal_filter_decoded_end_stream) {
+    ASSERT(!state_.decoder_filter_chain_complete_);
+    state_.decoder_filter_chain_complete_ = true;
+    // If the decoder filter chain was aborted (i.e. due to local reply)
+    // we rely on the encoding of end_stream to close the stream.
+    if (filter_manager_callbacks_.isHalfCloseEnabled() && !stopDecoderFilterChain()) {
+      checkAndCloseStreamIfFullyClosed();
+    }
+  }
+}
+
+void FilterManager::checkAndCloseStreamIfFullyClosed() {
+  ASSERT(filter_manager_callbacks_.isHalfCloseEnabled());
+  // When the independent half close is enabled the stream is always closed on error responses
+  // from the server.
+  if (filter_manager_callbacks_.responseHeaders().has_value()) {
+    const uint64_t response_status =
+        Http::Utility::getResponseStatus(filter_manager_callbacks_.responseHeaders().ref());
+    bool error_response =
```

Comment:
const, you can even initialize this in the if you want since it is not used elsewhere.

Reply:
Done